examples/docs/zh-CN/config/IPageConfig.md
## IPage 相关全局配置


### response 配置
response是关于isearch查询请求后，如何处理返回信息的相关配置
```js
config.set({
    response:{
        // 下表中的配置项
        data: 'data.records'
    }
})
```
|配置项|类型|说明|默认值|
|--|--|--|--|
|data|string|ITable需要的data数据|'data.records'|
|total|string|分页组件需要的total数据|'data.total'|
|beforeFunc|Function,null|自定义返回值处理函数，函数返回值必须是一个对象，包含两个属性：`data`,`total`|null|


### search 搜索分页相关参数
```js
config.set({
    search:{
        // 下表中的配置项
        mode:'offset'
    }
})
```
|配置项|类型|说明|默认值|
|--|--|--|--|
|mode|string|分页模式：page=页码模式，offset=偏移模式|'page'|
|pageSize|string|页码模式参数：分页尺寸|'pageSize'|
|pageNo|string|页码模式参数：页码|'pageNo'|
|limit|string|偏移模式参数：分页尺寸|'limit'|
|offset|string|偏移模式参数：(pageNo-1)*pageSize|'offset'|
|beforeFunc|Function,null|func(params)自定义查询前处理函数，函数返回值必须是一个对象，是给分页查询接口的分页参数。|null|

### ITable 配置
```js
config.set({
    // 下表中的配置项
    tableHeightMode:'data',
    dialogProps:{
        width: '50%',
        top:'3vh'
    }
})
```
|配置项|类型|说明|默认值|
|--|--|--|--|
|tableHeightMode|string|表格高度锁：仅在IPage组件中的ITable生效；parent=会自动撑满<父组件>，有滚动条;data=会显示全部数据撑开页面，不会有滚动条|'parent'|
|columnButtonProps|object|ITable 组件的默认传入数据，用来渲染表格操作栏，其中columnProps是传递给el-table-column组件的参数|{columnProps: {label: '操作',minWidth: 140,fixed: 'right',},}|


### el-pagination 组件配置
```js
config.set({
    paginationProps:{
        layout:"total, sizes, prev, pager, next, jumper"
    }
})
```
|配置项|类型|说明|默认值|
|--|--|--|--|
|tableHeightMode|string|表格高度锁：仅在IPage组件中的ITable生效；parent=会自动撑满<父组件>，有滚动条;data=会显示全部数据撑开页面，不会有滚动条|'parent'|
|paginationProps|object|传递给el-pagination分页组件的参数|{layout:"total, sizes, prev, pager, next, jumper"}|



### el-dialog 新增、修改弹窗配置
```js
config.set({
    // 下表中的配置项
    dialogProps:{
        width: '50%',
        top:'3vh'
    }
})
```
|配置项|类型|说明|默认值|
|--|--|--|--|
|dialogProps|object|弹窗默认的属性，将直接传递给IPage使用的增改数据el-dialog组件使用|{width: '50%',appendToBody: true}|
|dialogOn|object|弹窗默认的事件，将直接传递给IPage使用的增改数据el-dialog中|{}|

### toolbarButton 工具条按钮默认配置
```js
config.set({
    /** 默认toolbar上的RenderCell组件数据 */
    toolbarButton:{
        slot:'button',
        props:{
            size:'small',
            type:'text',
        }
    }
})
```

### button 默认表格操作区按钮配置
```js
config.set({
    /** 默认columnButtons里的RenderCell组件数据 */
    button:{
        slot: 'button',
        props: {
            size: 'small',
            type: 'text',
        },
    }
})
```

### editButton修改、deleteButton删除两个按钮的默认配置
它们两个按钮会先继承`button 默认表格操作区按钮配置`这个配置，然后再叠加自己各自的配置
```js
config.set({
    /** 默认columnButtons里的RenderCell组件数据 */
    editButton:{
        id: 'edit',
        tip: '编辑',
        /** 下面这部分会从button上继承
        slot: 'button',
        props: {
            size: 'small',
            type: 'text',
        },
        */
    },

    // 设置为false表示不需要这个按钮和功能
    deleteButton:false,
})
```
|配置项|类型|说明|默认值|
|--|--|--|--|
|editButton|object|bool=false|新增按钮的RenderCell组件数据，如果不需要编辑按钮，可以设置为false|-默认编辑-|
|deleteButton|object|bool=false|删除按钮的RenderCell组件数据，如果不需要删除按钮，可以设置为false|-默认删除-|


### addButton 新增按钮配置
用法如同editButton和deleteButton，不同之处在于addButton所在的位置是`toolbar`



examples/docs/zh-CN/config/GlobalConfig.md
## 全局配置

推荐直接在项目合适的位置创建一个`expConfig.js`的文件，然后在`main.js`中引入即可

### expConfig.js
```js
import {config} from "ipage-element-expand"
config.set({
    class:{
        /**IPage组件根节点附加class */
        IPageRoot:'ipage',
        /** ISearch组件根节点附加class */
        ISearchRoot:'isearch',
        /** ITable组件根节点附加class */
        ITableRoot:'itable',
        /** IForm组件根节点附加class
         *  以这个优先
         */
        IFormRoot:'iform',
        /** 操作区额外的class */
        IPageOperation:'ipage_operation', 
        /** 表格外层的class */
        tableWrap:'ipage_tableWrap', 
        /** 分页区额外的class */
        IPagePagination:'ipage_pagination',  
    },
    /// IPage 相关参数
    /** 返回值所在的地方 */
    response:{
        /** ITable需要的data数据 */
        data: 'data.records',
        /** 分页组件需要的total数据 */
        total: 'data.total',
        /** 自定义返回值处理函数
         * 函数返回值必须是一个对象，包含两个属性
         * data: 返回的数据
         * total: 返回的总数
        */
        beforeFunc: null,
    },
    /** 搜索分页相关参数 */
    search:{
        /** page=页码模式，offset=偏移模式 */
        mode:'page',
        /** 页码模式参数 */
        pageSize:'pageSize',
        pageNo:'pageNo',
        /** 偏移模式参数  */
        limit:'limit',
        offset:'offset',
        /** 自定义搜索参数处理函数
         * @param {object} params 当前的搜索参数
         * @returns {object} 返回一个对象，分页需要的参数
         */
        beforeFunc: null, 
    },
    
    /** 表格高度锁：仅在IPage组件中的ITable生效
     * parent=会自动撑满<父组件>，有滚动条
     * data=会显示全部数据撑开页面，不会有滚动条
     *  */
    tableHeightMode:'parent', // 
    /** 如果不是自动高度，将会撑满页面高度 */
    tableHeight:0,
    /** 弹窗默认的属性，将直接传递给IPage使用的增改数据el-dialog中 */
    dialogProps:{
        width: '50%',
        appendToBody: true,
    },
    /** 弹窗默认的事件，将直接传递给IPage使用的增改数据el-dialog中 */
    dialogOn:{ },
    
    /** ITable 组件的默认传入数据，用来渲染表格操作栏 */
    columnButtonProps:{
        columnProps: {
            label: '操作',
            minWidth: 140,
            fixed: 'right',
        },
    },
    /** 默认的el-pagination 组件的参数，直接传递给el-pagination */
    paginationProps:{
        layout:"total, sizes, prev, pager, next, jumper"
    },
    /** 默认toolbar上的RenderCell组件数据 */
    toolbarButton:{
        slot:'button',
        props:{
            size:'small',
            type:'text',
        }
    },
    /** 默认按钮的RenderCell组件数据 
     * 将用在addButton,deleteButton和默认的columnButtons里面
    */
    button:{
        slot: 'button',
        props: {
            size: 'small',
            type: 'text',
        },
    },
    /** 默认的addButton按钮RenderCell组件数据
     * 注意：IPage在使用addButton的时候，会再叠加一个on:{click:fn}的事件
     */
    addButton:{
        id: 'add',
        tip: '新增',
        props: {
            size: 'small',
            type: 'primary',
            icon: 'el-icon-plus',
        }
    },
    /** 默认的editButton按钮RenderCell组件数据
     * 注意：IPage在使用editButton的时候，会再叠加一个on:{click:fn}的事件
     */
    editButton:{
        id: 'edit',
        tip: '编辑',
    },
    /** 默认的deleteButton按钮RenderCell组件数据
     * 注意：IPage在使用deleteButton的时候，会再叠加一个on:{click:fn}的件
     */
    deleteButton:{
        id: 'delete',
        tip: '删除',
    },
    /** 默认的refreshButton按钮RenderCell组件数据
     * 注意：IPage在使用refreshButton的时候，会再叠加一个on:{click:fn}的件
     */
    refreshButton:{
        id:'refresh',
        props:{
            icon:'el-icon-refresh'
        },
    },
    /// IForm 相关参数
    /** 直接传递给IFrom组件内的el-form组件的class */
    formClass:"iform", // 旧版本参数，后续可能会删除
    /** 直接传递给IFrom组件内的el-form组件 */
    formProps:{
        "label-width": "120px"
    },
    /** 直接传递给IFrom组件内的el-form-item组件 */
    formOn:{ },
    /** IFrom组件的按钮区域RenderCell组件array */
    expandButtons:[],
    ////// 这部分参数暂时先做兼容处理，后续可能会删除，以后会变成RenderCel组件参数
    /** 是否显示IFrom组件的提交按钮 */
    showSubmitButton:true,
    submitTitle:"确定",
    cancelTitle:"取消",
    submitButtonProps:{
        type: "primary"
    },
    showCancelButton:false,
    cancelButtonProps:{
        type: "default"
    },
    /** 优化后的表单提交按钮 RenderCell 组件参数，或者false表示不显示 */
    formSubmitButton:{
        slot: 'button',
        props: {
            type: 'primary',
        }
    },
    /** 优化后的表单取消按钮 RenderCell 组件参数，或者false表示不显示 */
    formCancelButton:false,
    submitPreventRepeat:false,
    /// ISearch 相关参数
    /** ISearch组件内部的el-form参数 */
    searchFormProps:{
        inline:true
    },
    searchInput:{
        props:{
            size:'small',
            clearable:true,    
        }
    },
    searchButton:{
        props:{
            size:'small',   
        }
    },
    // 旧版本的参数，后续可能会删除，并替换成RenderCell组件参数
    showSearchButton:true,
    searchButtonProps:{
        type: "primary",
        icon: 'el-icon-search',
        size:"small"
    },
    showResetButton:true,
    resetButtonProps:{
        type: "default",
        icon: '',
        size:"small"
    },
    // 新版本参数
    searchSubmitButton:{
        slot: 'button',
        props: {
            icon: 'el-icon-search',
            type: 'primary',
            size:"small"
        }
    },
    searchResetButton:{
        slot: 'button',
        props: {
            type: 'default',
            size:"small"
        }
    },
    /** 强制所有搜索项目进more 比如在有些手机尺寸上可以用这个控制 */
    allinMore:false,
    /// ITable 相关参数
    tableProps:{},
    tableOn:{},
    /** 扩展渲染组件，请用jsx直接传入vnode */
    extendedRenderCell:{
        
    }
})
export default config
```

### main.js
```js
import expConfig from "@/utils/expConfig"
```


src/lib/config.js
import {deepAssign} from "./utils"
export const config = {
    options:{
        class:{
            /**IPage组件根节点附加class */
            IPageRoot:'ipage',
            /** ISearch组件根节点附加class */
            ISearchRoot:'isearch',
            /** ITable组件根节点附加class */
            ITableRoot:'itable',
            /** IForm组件根节点附加class
             *  以这个优先
             */
            IFormRoot:'iform',
            /** 操作区额外的class */
            IPageOperation:'ipage_operation', 
            /** 表格外层的class */
            tableWrap:'ipage_tableWrap', 
            /** 分页区额外的class */
            IPagePagination:'ipage_pagination',  
        },
        /// IPage 相关参数
        /** 返回值所在的地方 */
        response:{
            /** ITable需要的data数据 */
            data: 'data.records',
            /** 分页组件需要的total数据 */
            total: 'data.total',
            /** 自定义返回值处理函数
             * 函数返回值必须是一个对象，包含两个属性
             * data: 返回的数据
             * total: 返回的总数
            */
            beforeFunc: null,
        },
        /** 搜索分页相关参数 */
        search:{
            /** page=页码模式，offset=偏移模式 */
            mode:'page',
            /** 页码模式参数 */
            pageSize:'pageSize',
            pageNo:'pageNo',
            /** 偏移模式参数  */
            limit:'limit',
            offset:'offset',
            /** 自定义搜索参数处理函数
             * @param {object} params 当前的搜索参数
             * @returns {object} 返回一个对象，分页需要的参数
             */
            beforeFunc: null, 
        },
        
        /** 表格高度锁：仅在IPage组件中的ITable生效
         * parent=会自动撑满<父组件>，有滚动条
         * data=会显示全部数据撑开页面，不会有滚动条
         *  */
        tableHeightMode:'parent', // 
        /** 如果不是自动高度，将会撑满页面高度 */
        tableHeight:0,
        /** 弹窗默认的属性，将直接传递给IPage使用的增改数据el-dialog中 */
        dialogProps:{
            width: '50%',
            appendToBody: true,
        },
        /** 弹窗默认的事件，将直接传递给IPage使用的增改数据el-dialog中 */
        dialogOn:{ },
        
        /** ITable 组件的默认传入数据，用来渲染表格操作栏 */
        columnButtonProps:{
            columnProps: {
                label: '操作',
                minWidth: 140,
                fixed: 'right',
            },
        },
        /** 默认的el-pagination 组件的参数，直接传递给el-pagination */
        paginationProps:{
            layout:"total, sizes, prev, pager, next, jumper"
        },
        /** 默认toolbar上的RenderCell组件数据 */
        toolbarButton:{
            slot:'button',
            props:{
                size:'small',
                type:'text',
            }
        },
        /** 默认按钮的RenderCell组件数据 
         * 将用在addButton,deleteButton和默认的columnButtons里面
        */
        button:{
            slot: 'button',
            props: {
                size: 'small',
                type: 'text',
            },
        },
        /** 默认的addButton按钮RenderCell组件数据
         * 注意：IPage在使用addButton的时候，会再叠加一个on:{click:fn}的事件
         */
        addButton:{
            id: 'add',
            tip: '新增',
            props: {
                size: 'small',
                type: 'primary',
                icon: 'el-icon-plus',
            }
        },
        /** 默认的editButton按钮RenderCell组件数据
         * 注意：IPage在使用editButton的时候，会再叠加一个on:{click:fn}的事件
         */
        editButton:{
            id: 'edit',
            tip: '编辑',
        },
        /** 默认的deleteButton按钮RenderCell组件数据
         * 注意：IPage在使用deleteButton的时候，会再叠加一个on:{click:fn}的事件
         */
        deleteButton:{
            id: 'delete',
            tip: '删除',
        },
        /** 默认的refreshButton按钮RenderCell组件数据
         * 注意：IPage在使用refreshButton的时候，会再叠加一个on:{click:fn}的事件
         */
        refreshButton:{
            id:'refresh',
            props:{
                icon:'el-icon-refresh'
            },
        },
        /// IForm 相关参数
        /** 直接传递给IFrom组件内的el-form组件的class */
        formClass:"iform", // 旧版本参数，后续可能会删除
        /** 直接传递给IFrom组件内的el-form组件 */
        formProps:{
            "label-width": "120px"
        },
        /** 直接传递给IFrom组件内的el-form-item组件 */
        formOn:{ },
        /** IFrom组件的按钮区域RenderCell组件array */
        expandButtons:[],

        ////// 这部分参数暂时先做兼容处理，后续可能会删除，以后会变成RenderCell组件参数
        /** 是否显示IFrom组件的提交按钮 */
        showSubmitButton:true,
        submitTitle:"确定",
        cancelTitle:"取消",
        submitButtonProps:{
            type: "primary"
        },
        showCancelButton:false,
        cancelButtonProps:{
            type: "default"
        },
        /** 优化后的表单提交按钮 RenderCell 组件参数，或者false表示不显示 */
        formSubmitButton:{
            slot: 'button',
            props: {
                type: 'primary',
            }
        },
        /** 优化后的表单取消按钮 RenderCell 组件参数，或者false表示不显示 */
        formCancelButton:false,
        submitPreventRepeat:false,
        /// ISearch 相关参数
        /** ISearch组件内部的el-form参数 */
        searchFormProps:{
            inline:true
        },
        searchInput:{
            props:{
                size:'small',
                clearable:true,    
            }
        },
        searchButton:{
            props:{
                size:'small',   
            }
        },
        // 旧版本的参数，后续可能会删除，并替换成RenderCell组件参数
        showSearchButton:true,
        searchButtonProps:{
            type: "primary",
            icon: 'el-icon-search',
            size:"small"
        },
        showResetButton:true,
        resetButtonProps:{
            type: "default",
            icon: '',
            size:"small"
        },
        // 新版本参数
        searchSubmitButton:{
            slot: 'button',
            props: {
                icon: 'el-icon-search',
                type: 'primary',
                size:"small"
            }
        },
        searchResetButton:{
            slot: 'button',
            props: {
                type: 'default',
                size:"small"
            }
        },
        /** 强制所有搜索项目进more 比如在有些手机尺寸上可以用这个控制 */
        allinMore:false,
        /// ITable 相关参数
        tableProps:{},
        tableOn:{},
        /** 扩展渲染组件，请用jsx直接传入vnode */
        extendedRenderCell:{
            
        }
    },
    /** 设置参数，options可以传入参数的内容，保持正确的层级关系，会深度合并 */
    set(options) {
        this.options = deepAssign(this.options,options)
    },
    
    /** 获取配置信息，deepPath可以是一个string，比如
     * 'dialogProps.appendToBody' 它将获取this.options.dialogProps.appendToBody
     * 不传递的话，返回整个配置
     */
    get(deepPath) {
        if (deepPath) {
            const paths = deepPath.split('.');
            let result = this.options;
            for (let path of paths) {
                // todo 这里可能要处理异常
                result = result[path];
            }
            return result;
        } else {
            return this.options;
        }
    }
}